Move BitmapImage format conversion to a new class name ImageAdapter

shallawa opened the pull request:
#### ab92c87e9f99256f3390219e570a3dc1d0debee6
<pre>
Move BitmapImage format conversion to a new class name ImageAdapter
<a href="https://bugs.webkit.org/show_bug.cgi?id=267254">https://bugs.webkit.org/show_bug.cgi?id=267254</a>
<a href="https://rdar.apple.com/120690546">rdar://120690546</a>

Reviewed by Cameron McCormack.

The classes Image and BitmapImage are overloaded with many platform functions. The
platform-specific format conversion is not tied to BitmapImage or Image. It relies
mostly on the NativeImage. So move the format conversion to a new class named
ImageAdapter. Move also BitmapImage::nativeImageOfSize() and framesNativeImages()
to this new class since they are only used for image format conversion.

* Source/WebCore/Headers.cmake:
* Source/WebCore/PlatformWin.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/SourcesGTK.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/DataTransferMac.mm:
(WebCore::DataTransfer::createDragImage const):
* Source/WebCore/editing/cocoa/HTMLConverter.mm:
(fileWrapperForElement):
* Source/WebCore/loader/cache/CachedImage.cpp:
(WebCore::CachedImage::brokenImage const):
* Source/WebCore/platform/ScrollView.cpp:
(WebCore::ScrollView::paintPanScrollIcon):
* Source/WebCore/platform/audio/cocoa/MediaSessionManagerCocoa.mm:
(WebCore::MediaSessionManagerCocoa::setNowPlayingInfo):
* Source/WebCore/platform/cocoa/DragImageCocoa.mm:
(WebCore::createDragImageFromImage):
* Source/WebCore/platform/graphics/BitmapImage.cpp:
(WebCore::BitmapImage::~BitmapImage):
(WebCore::BitmapImage::destroyDecodedData):
(WebCore::BitmapImage::nativeImageAtIndexCacheIfNeeded):
(WebCore::BitmapImage::nativeImage):
(WebCore::BitmapImage::nativeImageForCurrentFrame):
(WebCore::BitmapImage::draw):
(WebCore::BitmapImage::drawPattern):
(WebCore::BitmapImage::decode):
(WebCore::BitmapImage::dump const):
(WebCore::BitmapImage::frameImageAtIndexCacheIfNeeded): Deleted.
(WebCore::BitmapImage::nativeImageOfSize): Deleted.
(WebCore::BitmapImage::framesNativeImages): Deleted.
* Source/WebCore/platform/graphics/BitmapImage.h:
* Source/WebCore/platform/graphics/Image.cpp:
(WebCore::Image::adapter):
(WebCore::Image::invalidateAdapter):
(WebCore::Image::fillWithSolidColor):
(WebCore::Image::drawTiled):
(WebCore::operator&lt;&lt;):
(WebCore::BitmapImage::invalidatePlatformData): Deleted.
(WebCore::Image::loadPlatformResource): Deleted.
* Source/WebCore/platform/graphics/Image.h:
(WebCore::Image::frameCount const):
(WebCore::Image::nativeImageAtIndex):
(WebCore::Image::nativeImageAtIndexCacheIfNeeded):
(WebCore::Image::nativeImageOfSize): Deleted.
(WebCore::Image::nsImage): Deleted.
(WebCore::Image::snapshotNSImage): Deleted.
(WebCore::Image::tiffRepresentation): Deleted.
(WebCore::Image::getHBITMAP): Deleted.
(WebCore::Image::getHBITMAPOfSize): Deleted.
(WebCore::Image::gdkPixbuf): Deleted.
(WebCore::Image::gdkTexture): Deleted.
(WebCore::Image::drawFrameMatchingSourceSize): Deleted.
* Source/WebCore/platform/graphics/ImageAdapter.cpp: Renamed from Source/WebCore/platform/graphics/win/ImageWin.cpp.
(WebCore::ImageAdapter::loadPlatformResource):
(WebCore::ImageAdapter::invalidate):
(WebCore::ImageAdapter::nativeImageOfSize):
(WebCore::ImageAdapter::allNativeImages):
* Source/WebCore/platform/graphics/ImageAdapter.h: Added.
(WebCore::ImageAdapter::ImageAdapter):
(WebCore::ImageAdapter::image const):
* Source/WebCore/platform/graphics/gtk/ImageAdapterGtk.cpp: Renamed from Source/WebCore/platform/graphics/gtk/ImageGtk.cpp.
(WebCore::ImageAdapter::loadPlatformResource):
(WebCore::ImageAdapter::invalidate):
(WebCore::ImageAdapter::gdkPixbuf):
(WebCore::ImageAdapter::gdkTexture):
* Source/WebCore/platform/graphics/mac/ImageAdapterMac.mm: Renamed from Source/WebCore/platform/graphics/mac/ImageMac.mm.
(WebCore::ImageAdapter::loadPlatformResource):
(WebCore::ImageAdapter::tiffRepresentation):
(WebCore::ImageAdapter::invalidate):
(WebCore::ImageAdapter::nsImage):
(WebCore::ImageAdapter::snapshotNSImage):
* Source/WebCore/platform/graphics/win/ImageAdapterWin.cpp: Renamed from Source/WebCore/platform/graphics/win/ImageCairoWin.cpp.
(WebCore::ImageAdapter::loadPlatformResource):
(WebCore::ImageAdapter::nativeImageOfHBITMAP):
(WebCore::ImageAdapter::invalidate):
(WebCore::ImageAdapter::getHBITMAP):
(WebCore::ImageAdapter::getHBITMAPOfSize):
* Source/WebCore/platform/gtk/CursorGtk.cpp:
(WebCore::createCustomCursor):
* Source/WebCore/platform/mac/CursorMac.mm:
(WebCore::createCustomCursor):
* Source/WebCore/platform/mac/PasteboardMac.mm:
(WebCore::Pasteboard::write):
* Source/WebCore/platform/win/CursorWin.cpp:
(WebCore::createSharedCursor):
(WebCore::loadCursorByName):
* Source/WebCore/platform/win/PasteboardWin.cpp:
(WebCore::Pasteboard::writeImage):
* Source/WebCore/rendering/RenderLayerScrollableArea.cpp:
(WebCore::RenderLayerScrollableArea::drawPlatformResizerImage):
* Source/WebCore/rendering/RenderThemeMac.mm:
(WebCore::createAttachmentPlaceholderImage):
* Source/WebKit/UIProcess/API/gtk/DragSourceGtk3.cpp:
(WebKit::DragSource::DragSource):
* Source/WebKit/UIProcess/API/gtk/DragSourceGtk4.cpp:
(WebKit::DragSource::begin):
* Source/WebKit/UIProcess/gtk/ClipboardGtk3.cpp:
(WebKit::Clipboard::write):
* Source/WebKit/UIProcess/gtk/ClipboardGtk4.cpp:
(WebKit::Clipboard::write):
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::provideDataForPasteboard):
* Source/WebKitLegacy/mac/DOM/DOM.mm:
(-[DOMElement image]):
(-[DOMElement _imageTIFFRepresentation]):
* Source/WebKitLegacy/mac/Misc/WebElementDictionary.mm:
(-[WebElementDictionary _image]):
* Source/WebKitLegacy/mac/WebCoreSupport/WebContextMenuClient.mm:
(WebContextMenuClient::imageForCurrentSharingServicePickerItem):
(WebContextMenuClient::contextMenuForEvent):
* Source/WebKitLegacy/mac/WebCoreSupport/WebFrameLoaderClient.mm:
(webGetNSImage):
* Source/WebKitLegacy/mac/WebView/WebHTMLView.mm:
(-[WebHTMLView pasteboard:provideDataForType:]):
* Tools/TestWebKitAPI/Tests/WebCore/win/BitmapImage.cpp:
(TestWebKitAPI::TEST):

Canonical link: <a href="https://commits.webkit.org/273772@main">https://commits.webkit.org/273772@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ad338a9c01738ccc6b97f0e79320ebcd37fbc108

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/36614 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/15554 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/38840 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/39275 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/32825 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/37848 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/18028 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/12633 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/31439 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/37175 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/13112 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/32384 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/11486 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/11491 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/32639 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/40519 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/33176 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/32987 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/37425 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/11749 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/9590 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/35537 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/13429 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8302 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/12166 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/12634 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

heycam commented on Source/WebCore/platform/graphics/ImageAdapter.h:
```diff
@@ -0,0 +1,114 @@
+    RefPtr<NativeImage> nativeImageOfSize(const IntSize&);
+    Vector<Ref<NativeImage>> allNativeImages();
+
+    Image& m_image;
```
What do you think about making this a `WeakRef`, for safety?

heycam commented on Tools/TestWebKitAPI/Tests/WebCore/win/BitmapImage.cpp:
```diff
@@ -1,61 +0,0 @@
-// Test that there is no crash when BitmapImage::getHBITMAPOfSize() is called
-// for an image with empty frames (BitmapImage::frameAtIndex(i) return null), WebKit Bug 102689.
-
-TEST(WebCore, BitmapImageEmptyFrameTest)
```
Should we preserve this test?

shallawa committed:
Move BitmapImage format conversion to a new class name ImageAdapter
https://bugs.webkit.org/show_bug.cgi?id=267254
rdar://120690546

Reviewed by Cameron McCormack.

The classes Image and BitmapImage are overloaded with many platform functions. The
platform-specific format conversion is not tied to BitmapImage or Image. It relies
mostly on the NativeImage. So move the format conversion to a new class named
ImageAdapter. Move also BitmapImage::nativeImageOfSize() and framesNativeImages()
to this new class since they are only used for image format conversion.

* Source/WebCore/Headers.cmake:
* Source/WebCore/PlatformWin.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/SourcesGTK.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/DataTransferMac.mm:
(WebCore::DataTransfer::createDragImage const):
* Source/WebCore/editing/cocoa/HTMLConverter.mm:
(fileWrapperForElement):
* Source/WebCore/loader/cache/CachedImage.cpp:
(WebCore::CachedImage::brokenImage const):
* Source/WebCore/platform/ScrollView.cpp:
(WebCore::ScrollView::paintPanScrollIcon):
* Source/WebCore/platform/audio/cocoa/MediaSessionManagerCocoa.mm:
(WebCore::MediaSessionManagerCocoa::setNowPlayingInfo):
* Source/WebCore/platform/cocoa/DragImageCocoa.mm:
(WebCore::createDragImageFromImage):
* Source/WebCore/platform/graphics/BitmapImage.cpp:
(WebCore::BitmapImage::~BitmapImage):
(WebCore::BitmapImage::destroyDecodedData):
(WebCore::BitmapImage::nativeImageAtIndexCacheIfNeeded):
(WebCore::BitmapImage::nativeImage):
(WebCore::BitmapImage::nativeImageForCurrentFrame):
(WebCore::BitmapImage::draw):
(WebCore::BitmapImage::drawPattern):
(WebCore::BitmapImage::decode):
(WebCore::BitmapImage::dump const):
(WebCore::BitmapImage::frameImageAtIndexCacheIfNeeded): Deleted.
(WebCore::BitmapImage::nativeImageOfSize): Deleted.
(WebCore::BitmapImage::framesNativeImages): Deleted.
* Source/WebCore/platform/graphics/BitmapImage.h:
* Source/WebCore/platform/graphics/Image.cpp:
(WebCore::Image::adapter):
(WebCore::Image::invalidateAdapter):
(WebCore::Image::fillWithSolidColor):
(WebCore::Image::drawTiled):
(WebCore::operator<<):
(WebCore::BitmapImage::invalidatePlatformData): Deleted.
(WebCore::Image::loadPlatformResource): Deleted.
* Source/WebCore/platform/graphics/Image.h:
(WebCore::Image::frameCount const):
(WebCore::Image::nativeImageAtIndex):
(WebCore::Image::nativeImageAtIndexCacheIfNeeded):
(WebCore::Image::nativeImageOfSize): Deleted.
(WebCore::Image::nsImage): Deleted.
(WebCore::Image::snapshotNSImage): Deleted.
(WebCore::Image::tiffRepresentation): Deleted.
(WebCore::Image::getHBITMAP): Deleted.
(WebCore::Image::getHBITMAPOfSize): Deleted.
(WebCore::Image::gdkPixbuf): Deleted.
(WebCore::Image::gdkTexture): Deleted.
(WebCore::Image::drawFrameMatchingSourceSize): Deleted.
* Source/WebCore/platform/graphics/ImageAdapter.cpp: Renamed from Source/WebCore/platform/graphics/win/ImageWin.cpp.
(WebCore::ImageAdapter::loadPlatformResource):
(WebCore::ImageAdapter::invalidate):
(WebCore::ImageAdapter::nativeImageOfSize):
(WebCore::ImageAdapter::allNativeImages):
* Source/WebCore/platform/graphics/ImageAdapter.h: Added.
(WebCore::ImageAdapter::ImageAdapter):
(WebCore::ImageAdapter::image const):
* Source/WebCore/platform/graphics/gtk/ImageAdapterGtk.cpp: Renamed from Source/WebCore/platform/graphics/gtk/ImageGtk.cpp.
(WebCore::ImageAdapter::loadPlatformResource):
(WebCore::ImageAdapter::invalidate):
(WebCore::ImageAdapter::gdkPixbuf):
(WebCore::ImageAdapter::gdkTexture):
* Source/WebCore/platform/graphics/mac/ImageAdapterMac.mm: Renamed from Source/WebCore/platform/graphics/mac/ImageMac.mm.
(WebCore::ImageAdapter::loadPlatformResource):
(WebCore::ImageAdapter::tiffRepresentation):
(WebCore::ImageAdapter::invalidate):
(WebCore::ImageAdapter::nsImage):
(WebCore::ImageAdapter::snapshotNSImage):
* Source/WebCore/platform/graphics/win/ImageAdapterWin.cpp: Renamed from Source/WebCore/platform/graphics/win/ImageCairoWin.cpp.
(WebCore::ImageAdapter::loadPlatformResource):
(WebCore::ImageAdapter::nativeImageOfHBITMAP):
(WebCore::ImageAdapter::invalidate):
(WebCore::ImageAdapter::getHBITMAP):
(WebCore::ImageAdapter::getHBITMAPOfSize):
* Source/WebCore/platform/gtk/CursorGtk.cpp:
(WebCore::createCustomCursor):
* Source/WebCore/platform/mac/CursorMac.mm:
(WebCore::createCustomCursor):
* Source/WebCore/platform/mac/PasteboardMac.mm:
(WebCore::Pasteboard::write):
* Source/WebCore/platform/win/CursorWin.cpp:
(WebCore::createSharedCursor):
(WebCore::loadCursorByName):
* Source/WebCore/platform/win/PasteboardWin.cpp:
(WebCore::Pasteboard::writeImage):
* Source/WebCore/rendering/RenderLayerScrollableArea.cpp:
(WebCore::RenderLayerScrollableArea::drawPlatformResizerImage):
* Source/WebCore/rendering/RenderThemeMac.mm:
(WebCore::createAttachmentPlaceholderImage):
* Source/WebKit/UIProcess/API/gtk/DragSourceGtk3.cpp:
(WebKit::DragSource::DragSource):
* Source/WebKit/UIProcess/API/gtk/DragSourceGtk4.cpp:
(WebKit::DragSource::begin):
* Source/WebKit/UIProcess/gtk/ClipboardGtk3.cpp:
(WebKit::Clipboard::write):
* Source/WebKit/UIProcess/gtk/ClipboardGtk4.cpp:
(WebKit::Clipboard::write):
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::provideDataForPasteboard):
* Source/WebKitLegacy/mac/DOM/DOM.mm:
(-[DOMElement image]):
(-[DOMElement _imageTIFFRepresentation]):
* Source/WebKitLegacy/mac/Misc/WebElementDictionary.mm:
(-[WebElementDictionary _image]):
* Source/WebKitLegacy/mac/WebCoreSupport/WebContextMenuClient.mm:
(WebContextMenuClient::imageForCurrentSharingServicePickerItem):
(WebContextMenuClient::contextMenuForEvent):
* Source/WebKitLegacy/mac/WebCoreSupport/WebFrameLoaderClient.mm:
(webGetNSImage):
* Source/WebKitLegacy/mac/WebView/WebHTMLView.mm:
(-[WebHTMLView pasteboard:provideDataForType:]):
* Tools/TestWebKitAPI/Tests/WebCore/win/BitmapImage.cpp:
(TestWebKitAPI::TEST):

Canonical link: https://commits.webkit.org/273772@main
